pkgs/ok_http/lib/src/ok_http_web_socket.dart

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:jni/jni.dart';
+import 'package:web_socket/web_socket.dart';
+
+import 'jni/bindings.dart' as bindings;
+import 'ok_http_client.dart';
+
+/// A [WebSocket] implemented using the OkHttp library's
+/// [WebSocket](https://square.github.io/okhttp/5.x/okhttp/okhttp3/-web-socket/index.html)
+/// API.
+class OkHttpWebSocket implements WebSocket {
+  late bindings.OkHttpClient _client;
+  late final bindings.WebSocket _webSocket;
+  final _events = StreamController<WebSocketEvent>();
+  String? _protocol;
+
+  OkHttpWebSocket._() {
+    // Add the WSInterceptor to prevent response parsing errors.
+    _client = bindings.WSInterceptor.Companion
+        .addWSInterceptor(bindings.OkHttpClient_Builder())
+        .build();
+  }
+
+  /// Create a new WebSocket connection using `OkHttp`'s
+  /// [WebSocket](https://square.github.io/okhttp/5.x/okhttp/okhttp3/-web-socket/index.html)
+  /// API.
+  ///
+  /// The URL supplied in [url] must use the scheme ws or wss.
+  ///
+  /// If provided, the [protocols] argument indicates that subprotocols that
+  /// the peer is able to select. See
+  /// [RFC-6455 1.9](https://datatracker.ietf.org/doc/html/rfc6455#section-1.9).
+  static Future<WebSocket> connect(Uri url,
+          {Iterable<dynamic>? protocols}) async =>
+      OkHttpWebSocket._()._connect(url, protocols);
+
+  Future<WebSocket> _connect(Uri url, Iterable<dynamic>? protocols) async {
+    if (!url.isScheme('ws') && !url.isScheme('wss')) {
```

Comment:
Is there a reason why you created an empty constructor and do the work in the `_connect` method instead of having the same logic in `connect`?

I'm asking because it might be since if the approach is the same as is used by [`BrowserWebSocket`](https://github.com/dart-lang/http/blob/4178b67e7970ff02926e79649db5ab473b46cf30/pkgs/web_socket/lib/src/browser_web_socket.dart#L28)

Reply:
Yes! Since we need an instance of `bindings.OkHttpClient` to both open and close WebSockets, we cannot access it statically inside `connect()`, which I had to use to make the implementation more conformant with the other WebSockets.

Inside `static connect()`, we make a new instance of the OkHttpClient and then create a non-static `_connect` method that can access this instance. 

(I added some docs to explain the gist of it)